### ISSUE TITLE
How is the concatenation operation performed in the model?

### ISSUE THREAD

**Issue by hmartelb** [twitter-research/sign#1]
Hi, thank you for sharing your work! 

I am trying to understand the implementation from the paper and I came across a question about the concatenation operation. I assume that the MLP class from the code represents the architecture from Figure 1. 

![Fig1](https://user-images.githubusercontent.com/25375406/100522045-c6cfcb80-319f-11eb-9b43-b84e4a0a667f.PNG)

The forward pass in the code, however, does not explicitly contain a concatenate operation. Looking at this loop it seems that the layers are being applied in cascade instead of forming parallel branches.

```python
    def forward(self, x):
        for i, lin in enumerate(self.lins[:-1]):
            x = lin(x)
            x = self.bns[i](x)
            x = F.relu(x)
            x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.lins[-1](x)
        return torch.log_softmax(x, dim=-1)
```

In other words, what I would expect is to have something similar to this (I have not tested the code):

```python
    def forward(self, x):
        branches = []
        for i, lin in enumerate(self.lins[:-1]):
            _x = lin(x)
            _x = self.bns[i](_x)
            _x = F.relu(_x)
            _x = F.dropout(_x, p=self.dropout, training=self.training)
           branches.append(_x)
        x = torch.cat(branches)
        x = self.lins[-1](x)
        return torch.log_softmax(x, dim=-1)
```

Can you indicate me whether or not this interpretation is correct? Am I missing something and the concatenation takes place in some other part of the code? Thank you in advance.

**Comment by Noired** [comment]
Hi!

Your interpretation is indeed correct, in that the code we reported implements a slightly different -- and somewhat more expressive --  version of SIGN. Here, node features downstream the application of the graph filtering operations are concatenated _before_ being projected (see [line 125](https://github.com/twitter-research/sign/blob/master/sign_training.py#L125) of file 'sign_training.py'). This version could, in principle, simulate the architecture depicted in Figure 1 of the paper by learning a single block-diagonal parameter matrix **W** where each block corresponds to matrices **Theta**_0, ..., **Theta**_r.

For completeness, we will consider sharing the code implementing the exact architecture in Figure 1, although of trivial implementation.

Hope this helps!

**Comment by hmartelb** [comment]
Understood, this answer is definitely helpful. Thank you again! 
I am closing this issue.
